### ISSUE TITLE
Meta tracking issue for branch hints (RFC 1131)

### ISSUE THREAD

**Issue by nikomatsakis** [rust-lang/rust#26179]
Tracking and meta discussion for branch and path hinting, including:

- `likely`/`unlikely` as accepted in rust-lang/rfcs#1131 (much has changed since this RFC)
- `core::hint::{likely, unlikely}` https://github.com/rust-lang/rust/issues/151619
- `core::hint::cold_path` https://github.com/rust-lang/rust/issues/136873
- `bool::select_unpredictable` https://github.com/rust-lang/rust/issues/133962
- `#[cold]`:
  - https://github.com/rust-lang/rust/pull/120193

**Comment by pythonesque** [comment]
FWIW, looking through my own code, I actually think the fact that this can't apply to match variants is going to be fairly hard on idiomatic Rust (for example, it means you can't use it when matching on an `Option`).

**Comment by Aatch** [comment]
@pythonesque once I get this done, I plan on writing a follow-up RFC for a `#[likely]` attribute you can put on match arms. Since it's already possible to put attributes on arms (to support conditional compilation of them), it should be fairly easy to implement.

**Comment by Craig-Macomber** [comment]
It might be worth having an "unpredictable" option in addition to likely and unlikely. There is now something to lower it to in LLVM: http://reviews.llvm.org/D12341

**Comment by apasel422** [comment]
@Aatch Are you still working on this?

**Comment by vi** [comment]
Shall `#[likely]` be also on Result?

```
enum Result<T, E> {

   Ok(T),
   Err(E)
}
```

**Comment by Aatch** [comment]
@vi I think hinting should be only be used rarely, so it being on the type doesn't seem like a good idea. `Result` is a bit more general purpose than just error-handling, sot that kind of hinting would impact code where it's inappropriate. Hinting should only really be done at the site itself.

**Comment by vi** [comment]
@Aatch, Is it a good idea to have `likely` in every place where something is likely. Can it also serve for self-documentation: follow "likelies" and see happy path.

**Comment by Aatch** [comment]
@vi is is not. As a general rule, people are pretty bad at knowing what cases are *actually* likely or not. Branch hinting is a fairly specific optimisation hint, used incorrectly it can actually make code slower. It should also only really be used when the unlikely case is known to be very unlikely.

Basically, you should only use it if it makes your code faster, otherwise it should be there.

**Comment by pczarn** [comment]
@vi It might be good idea if the likely thing is cheap, and the unlikely thing is expensive. For example, `Vec::push` is expensive if it causes a reallocation, and very cheap otherwise.

**Comment by Stebalien** [comment]
Instead of arguing about it, why not just test it? Annotate `Result`, and then see if rustc bootstraps significantly faster. This won't give a definitive answer but will help determine if it's worth considering.

**Comment by comex** [comment]
Now that MIR is a thing, it would be nice if this were implemented.

**Comment by nikomatsakis** [comment]
@comex I do agree =) 

@Aatch Not sure what your time is like, but if you don't have time to implement this, it seems like a case where we might be able to mentor someone. The idea would be to leave a comment explaining roughly how you think it should work, and then tagging the issue with `E-mentor` and `E-help-wanted`. I of course can do this too and will try to remember to do so. =)

**Comment by nikomatsakis** [comment]
Unassigning @Aatch since, empirically, he doesn't seem to be actively hacking on this right now. =) (Feel free to correct me.)

**Comment by seanmonstar** [comment]
@nikomatsakis PR up, though I could use some help adding tests.

**Comment by Aatch** [comment]
Copying @nagisa's comment from #36181:

> #llvm told me expect intrinsic is more harmful than it is helpful and that the branch weight metadata should be emitted instead. Might be worth to go back and reconsider the RFC, since the branch weight metadata is more powerful and expressive and would be better expressed as attributes in source too.

>> <nagisa> why is clang > 3.6 not emiting expect intrinsic for __builtin_expect?
>> <zygoloid> nagisa: it turns out that the expect intrinsic overall harmed the optimizer by inhibiting certain optimizations
>> <zygoloid> so we now directly lower __builtin_expect to branch probabilities in the frontend when generating IR

I think that keeping the intrinsics is a good idea. They should be fairly obvious for people coming from C/C++. Attributes aren't mutually exclusive with the intrinsics, so we're not exactly cutting ourselves off with intrinsics.

Supporting branch weights at the MIR level would probably be the easiest implementation for lowering to the metadata ourselves.

**Comment by chriskrycho** [comment]
Just adding a note here as I'm processing everything in #38643: it will need docs before stabilization. @nikomatsakis would you mind adding a note to that effect in the root of the issue?

**Comment by nikomatsakis** [comment]
@chriskrycho done

**Comment by bluss** [comment]
Will these intrinsics be stabilized as they are? Under which path?

**Comment by seanmonstar** [comment]
Every few months, because I've forgotten the results, I try these out in httparse, on branches I'm absolutely certain are always taken, and see no gain, with a wild range in variance when running the becnhmarks.

I'm not adept at all with either assembly or llvm-ir when compiling with optimizations to tell how they are implemented.

There's older comments that the expect intrinsic in LLVM is harmful, that it should be done with branch weights instead. But when I go diving in LLVMs source now, it appears there is a pass that already lowers expects into branch weights anyways?

**Comment by nagisa** [comment]
The reason they're harmful is because they're still very opaque calls all
the way till they're lowered, inhibiting all the early optimisations which
are not specifically handling these intrinsics. Just don't use them.

**Comment by hsivonen** [comment]
> The reason they're harmful is because they're still very opaque calls all the way till they're lowered, inhibiting all the early optimisations which are not specifically handling these intrinsics. Just don't use them.

An earlier comment says (if I'm reading it right) that clang deals with the corresponding C intrinsic by generating branch probabilities in the IR instead of IR intrinsic.

Are there technical reasons against doing the same in rustc?

**Comment by nagisa** [comment]
No reasons against generating branch probabilities in rustc, however it is not as trivial to implement it that way.

One could as well just go ahead and add the attributes that are more general *and* easier to implement instead.

**Comment by nikomatsakis** [comment]
I have to say, I am personally inclined to just remove this intrinsic. The results seem to be uneven at best and I feel no strong urgency to jump on this.

**Comment by ArtemGr** [comment]
> The results seem to be uneven at best

You mean profile-guided optimization is a hoax?

**Comment by GabrielMajeri** [comment]
According to the [Intel optimization manual](https://www.intel.com/content/dam/www/public/us/en/documents/manuals/64-ia-32-architectures-optimization-manual.pdf), the order in which branches are laid out in assembly does matter (likely code should come first, unlikely comes after).

However, the question is if LLVM isn't already smart enough to reorder stuff by itself. These intrinsics were added a long time ago, when compilers were bad at predicting stuff.

From what I've seen, if you use an `unwrap` or check for an error code (e.g. right before `return Err`), the compiler is smart enough to optimize the code and detect the more likely branch.

Even the Linux kernel has an [option](https://cateee.net/lkddb/web-lkddb/OPTIMIZE_INLINING.html) to allow the compiler to un-inline certain functions because newer versions of GCC are much better at inlining. Same thing could happen with `likely` / `unlikely`: if the compiler is smart enough, there is no need for the intrinsic.

TL; DR: these intrinsics are useful for people who love premature optimization, believing they can outsmart the compiler.

As a workaround, you can simply write your code such that the generated assembly is laid out as you want. Or use [PGO](https://github.com/vadimcn/cargo-pgo).

**Comment by vi** [comment]
> TL; DR: these intrinsics are useful for people who love premature optimization, believing they can outsmart the compiler.

Annotating the code with likely/unlikely can also help readability. What is marked likely for the compiler is also interpreted as likely by the reader, who can follow the "main path" in the code more easily.

**Comment by briansmith** [comment]
> @pythonesque once I get this done, I plan on writing a follow-up RFC for a #[likely] attribute you can put on match arms. Since it's already possible to put attributes on arms (to support conditional compilation of them), it should be fairly easy to implement.

We could perhaps just allow `#[cold]` and/or `#[inline(never)]` on blocks/statements instead of just functions:

```rust
let x = if condition {
    ....
} else #[cold] {
   ....
};

let y = match condition2 {
    Option(x) => x,
    None => #[cold] 1
};
```

Regarding likely/unlikely specifically, I think people are expecting that they'd only be used inside `if` expressions, but I don't see that restriction in the RFC.

Anyway, in lots of the code I write, regarding parsing and cryptography, we really do always want the "no problems" path to be fast. Not only do we generally not care about the performance of the unlikely "oops, that's not how it's supposed to be" path, but we in fact would like to have the compiler actively de-optimize that path, moving it far away from the fast path, to improve instruction cache locality.

Also, PGO sounds nice but for many applications, like most of mine, it would be a huge waste. Its hard enough to get good code coverage; it's nearly impossible to develop PGO-focused coverage that emphasizes the code paths for which one most cares about performance.

**Comment by hsivonen** [comment]
> TL; DR: these intrinsics are useful for people who love premature optimization, believing they can outsmart the compiler.

Possibly. And possibly these don't generate the best LLVM IR for the purpose given how LLVM consumes IR today.

However, I think it's not productive to suggest that the compiler always outsmarts the programmer and, therefore, there's no need to provide detailed control as an option. I think the appropriate use for intrinsics like these is being able to persist manual PGO (iterating `cargo bench` with different branch prediction hints) in the source code without requiring the user of the crate to set up an appropriate benchmark for compile-time profiling.

Last week, I had a `match` with consecutive (both in source order and in numerical order of the condition) identical arms and combining the identical match arms destroyed about 100 MB/s of throughput (from 9xx MB/s to 8xx MB/s), because the compiler put the checks in a different order. By rewriting the code as manually-ordered `if`s, I was able to do better than the original `match`.

**Comment by briansmith** [comment]
Even in the standard library, there are quite a few places where these kinds of improvements to optimization hints would be useful. We shouldn't need to factor out code into a separate function just to mark it unlikely/cold like in https://github.com/rust-lang/rust/commit/257bff3192e2c7313a4f8cfcac8839a573b42f6b#diff-5da4ecf5ad221f68bde0ae13f2a7c22dR700, for example.

Also, in my code, I generally want the compiler to optimize, to an extreme extent, `likely()` paths for speed and cache locality. Similarly, I want it to optimize, to an extreme extent, `unlikely()`/`#[cold]` paths for size and out-of-the-way cache locality. Also, like Henri said, sometimes predictable performance is very important and sometimes the profile-guided optimization has been done by hand to ensure such predictable performance. Hints like `likely()`/`unlikely()`/`#[cold]` are the only way we have to transfer the knowledge from the profiling phase to the codegen phase in those situations.

**Comment by ArtemGr** [comment]
> In general, you should prefer to use actual profile feedback for this (-fprofile-arcs), as programmers are notoriously bad at predicting how their programs actually perform.

Yes, it's a good idea to profile and *then* optimize.
If the compiler can help with automatic PGO, all the better.

Thing is, not having a way to pass branch probabilities information to LLVM puts Rust at a disadvantage over the C versions of the same algorithms that has been profiled and optimized.
On the other hand, having this implemented on the way of intrinsics will likely pave the way for the automatic PGO as well.

Mozilla relies on PGO to speed up their Firefox builds, for example, and they have published notable gains. Being able to do the PGO in Rust might help it compete with established performance scores.

**Comment by ArtemGr** [comment]
> In general, you should prefer to use actual profile feedback for this (-fprofile-arcs), as programmers are notoriously bad at predicting how their programs actually perform.

Yes, it's a good idea to profile and *then* optimize.
If the compiler can help with automatic PGO, all the better.

Thing is, not having a way to pass branch probabilities information to LLVM puts Rust at a disadvantage over the C versions of the same algorithms that has been profiled and optimized.

On the other hand, I wonder if having this implemented in the way of intrinsics would pave the way for the automatic PGO as well.
Or are we happy with doing PGO through clang, like there https://github.com/Geal/pgo-rust ?

Mozilla relies on PGO to speed up their Firefox builds, for example, and they have published notable gains. Being able to do the PGO in Rust might help it compete with established performance scores.

**Comment by nikomatsakis** [comment]
Has anyone successfully used these intrinsics to achieve speedups?

I prefer the `#[cold]` formulation, myself.

I didn't mean to suggest that there is no role for user-hinting, but merely that this particular formulation didn't seem like a big success. The LLVM developers (from what I understand) don't recommend using these instrinsics, and they seem to be usable in confusing ways (e.g., outside of an `if`).

**Comment by ArtemGr** [comment]
> Has anyone successfully used these intrinsics to achieve speedups?

@nikomatsakis, in the way they're implemented the intrinsics aren't very useful, but this can be improved by "generating branch probabilities in the IR", like clang does.

Using likely and unlikely we can achieve the speedups in C, so why shouldn't we be able to do this in Rust, given that the intrinsics are already there? Why not simply fix them to be useful?

**Comment by joshlf** [comment]
On the subject of whether they're useful: I've used them myself (though I've not benchmarked enough to tell whether they made a difference). One thing I did notice was that `rustc` doesn't appear to be able to do constant propagation through them because I had a call to `unlikely(<expr>)` where `<expr>` could easily be constant-propagated to `false` at compile-time, but the corresponding `if` branch was still included in the compiled output.

**Comment by hsivonen** [comment]
> Has anyone successfully used these intrinsics to achieve speedups?

On most of my attempts, not, but that might have been due to wrapping them in inlines, which [according to another issue doesn't work](https://github.com/rust-lang/rust/issues/45440). 

Today, however, I finally found a case where I saw a notable boost. If I add `unlikely` to [this `if`](https://github.com/hsivonen/encoding_rs/blob/cc902eaebcca81c176dc74c4686b86a5903a8b90/src/mem.rs#L512), I get a 33% performance improvement on Russian plain text. (In case anyone thinks that the function as a whole is an unnecessary optimization, it's notably faster than iterating over a `str` by `char` and calling `is_char_bidi()` for each one.)

Also: Why are `likely()` and `unlikely()` currently `unsafe`? As far as their non-performance semantics go, they just return their argument, which is safe, and modifying performance is safe, too.

**Comment by nikomatsakis** [comment]
> On most of my attempts, not, but that might have been due to wrapping them in inlines, which according to another issue doesn't work. .. Also: Why are likely() and unlikely() currently unsafe? As far as their non-performance semantics go, they just return their argument, which is safe, and modifying performance is safe, too.

Not good reasons. I think it's because of the (imo) rather artificial distinction between "intrinsics" and "lang items" in the compiler. I've been meaning forever to write an RFC collapsing the distinction -- basically just having "something well known to the compiler". But for whatever reason we have a distinction now, and intrinsic functions are (a) always unsafe and (b) never exposed directly (hence the inlining).

> Why not simply fix them to be useful?

My feeling was that since they are not working that well now (and are not in any particular use) and the because the interface is Not Great, maybe we should rip it out and try something better.

The reasons that I think the interface is not great are roughly that a function doesn't seem like the optimal choice here:

1. Apparently we have to do some mapping to something else for LLVM to make good use of them
2. What is the meaning of `let x = likely(...); if x { .. .}`?  Is that the same as `if likely(...) { ... }`? 
    - Tracking where the call to `likely` occurred is kind of painful around MIR too
    - Why use an "identity function" (which can be invoked from many places) when we are really trying to indicate what code is cold and which is not?

etc

So, in short, I don't have a strong opinion, just looking for opportunities to cull underutilized and not especially succesful language features and replace with something better.

**Comment by ArtemGr** [comment]
> My feeling was that since they are not working that well now (and are not in any particular use) and the because the interface is Not Great, maybe we should rip it out and try something better.

Good point, @nikomatsakis !

It might take some time though for a different branch hinting approach to come into the language.

I think that in Rust we have a strong precedence of having practical, but Not Great, solutions to be implemented in the language in the _unstable_ form. It allows us to have our pie and eat it too.

As of now Rust might be at some disadvantage over C/C++ in regards of branch hinting. There are existing algorithms and programs that have been optimized in C/C++ to use branch hinting effectively, and currently it might be problematic to port them to Rust while maintaining the same level of performance, because we lack a branch hinting solution of a quality comparable to GCC and CLANG implementations.

Implementing the likely/unlikely intrinsic as an unstable feature, till we have a better one, might be good for language adoption. It might also help those members of the community who want to port their code to a safer language but are hindered by performance constraints.

I think that a lot of us flock to Rust because it offers performance. It's one of the trifecta on the https://www.rust-lang.org/ page. Personally, I started to believe in Rust as a performance language when I've seen the likely/unlikely RFC merged. It pains me to see that Rust still isn't there.

Can I do something to convince you that we need something working _until_ we have something better?

**Comment by bluss** [comment]
@ArtemGr You already have it as  unstable feature: [`std::intrinsics::unlikely`](https://doc.rust-lang.org/nightly/std/intrinsics/fn.unlikely.html). Using it and evaluating the current implementation is already possible. As you can see the discussion here is already in relation to the existing implementation and its flaws.

**Comment by ArtemGr** [comment]
@bluss , it's not currently on par with CLANG / GCC implementations.

**Comment by gnzlbg** [comment]
From [P0479r2](http://www.open-std.org/jtc1/sc22/wg21/docs/papers/2017/p0479r2.html):

* It seems that `likely`/`unlikely` are particularly widely used in systems programming (linux kernel and web browsers: Firefox and Chrome). 
* the optimization hints can be used to improve/degrade performance up to 40% on synthetic examples (two examples are studied), and to a much lesser extent on real projects (two open-source projects are studied) by generating `cmp/jmp` for predictable branches instead of `cmov` instructions in x86 (PowerPC has instructions to override dynamic branch prediction, so this has an impact there as well).
* these intrinsics are not easy to use, but they "are not meant to be added to code without measuring the impact to ensure they do not end up degrading performance".

Coming back to Rust, I do think that we need something like `likely/unlikely`, but I don't like the way the intrinsics are currently implemented:

- they only apply to `if`-statements, there is no support for `if let`, `match`, ... (forcing users to rewrite a match into a sequence of `if`-statements just so that they can use `likely` in an arm to hint that the particular arm will be taken 99% of the time seems unnecessary)
- they make the expression "opaque" kind of like `test::black_box`, this looks unnecessary
- we already have the `#[cold]` attribute to indicate that a function is unlikely to be called, having a separate `unlikely` intrinsic/attribute might be unnecessary (if not, what is the difference?)

Basically what I want is a way to override the optimizers default of using `cmov`, by stating "this branch is predictable and is going to be/not be taken often" that works on all types of branches (match, if let, ...). I guess that for this to work it would need to be an attribute, and also, be passed to LLVM as branch prediction hints, which is what they are actually are.

**Comment by steveklabnik** [comment]
So, the C++ comittee is talking about standardizing `likely`/`unlikely`, and herb sutter had this to say: https://herbsutter.com/2018/04/02/trip-report-winter-iso-c-standards-meeting-jacksonville/

>  For example, for some code it is essential to make the uncommon path faster so that it is faster when you do hit it — and in such cases you would actually use this attribute in the inverted sense, so please also write a comment nearby to document that intent otherwise some hapless reader (more than likely your future self) will be temporarily confused.

[Some people on Reddit](https://www.reddit.com/r/cpp/comments/8980ww/trip_report_winter_iso_c_standards_meeting/dwqq0c1/) have ideas for different names that don't have this drawback.

**Comment by gnzlbg** [comment]
@steveklabnik Thanks for pointing that out! 

The current blocker here is that the current API is broken beyond all repair. 

I personally would prefer just allowing `#[cold]` in more situations to mean unlikely, and adding a `#[hot]` attribute that's the opposite of `#[cold]` to mean likely. But right now, any solution, even one with a bad name, would be infinitely better than what we have. Adding `#[likely]` and `#[unlikely]` attributes to unstable Rust would probably be the right first step to gain implementation experience without having to mess with the meaning of `#[cold]`, and whether these can be unified with `hot`/`cold` or renamed to something else is something that can be done afterwards.

**Comment by Amanieu** [comment]
I get a consistent [10-15% speedup](https://github.com/Amanieu/hashbrown/blob/master/src/raw/mod.rs#L11) in `hashbrown` (a fast hash table implementation) using these intrinsics.

**Comment by gnzlbg** [comment]
@Amanieu the problem is that one cannot use these intrinsics in `match`, `if-let`, ... statements, so we need something better here.

I think enough proof / motivation has already been given to convince most that these are useful, and P0479r2 contains a thorough performance analysis of properly and improperly using these intrinsics. 

As mentioned I wish we could just use `#[hot]` and `#[cold]` for this (two attributes that already exist):

```rust
if foo() { #![hot] } else { #![cold] }
if let Some(v) = opt { #![cold] }
match opt {
    Some(v) => { #![hot] ... },
    None => { #![cold] ... }.
}
```

but I don't think that it is easy to lower these to an `llvm.expect` intrinsic on the value that's responsible for the branch landing there.

**Comment by hsivonen** [comment]
> The current blocker here is that the current API is broken beyond all repair.

That's a bit of an exaggeration. The current design does work for plain `if` if changed technically from intrinsics to lang items to that `unsafe` isn't needed by policy.

Also, `#![hot]` and `#![cold]` have a different problem: They are at least confusing in the case of a chained `if`, which you'd have to rewrite as nested `if`s instead.

**Comment by gnzlbg** [comment]
> That's a bit of an exaggeration. 

If you want to apply this to `match`, `if let`, `while let`, etc., you'd have to do "something like this":

```rust
match foo {
   A => ...,
   core::intrinsic::likely(B) => ...,
   core::intrinsic::likely(C if ...) => ...,
   C => ...,
   _ => ...,
}
```

where `B` and `C if ...` are the pattern you are matching and `core::intrinsic::likely` is something that has to somehow disappear at the syntax level. Even for simple `if-else` code one cannot easily annotate the `else` branch using intrinsics:

```rust
if foo { 
   ...
} else if ...  { ... }
} else if ...  { ... }
} else if ...  { ... }
} core::intrinsics::unlikely(else) {  // ??
   ...
}
``` 

The `core::intrinsics::{likely,unlikely}` approach is definitely good enough as unstable stop gap until we get something better.  The performance consequences of these annotations are very hard to predict correctly, and we want people to use them only after appropriately evaluating that they get a positive effect. A solution that requires people to transform all of their conditional code into a series of `if ... { ... } else { ... }` statements just to be able to apply these intrinsics and test their effects is not IMO something worth stabilizing. This is what I meant with "the intrinsics approach is broken beyond all repair".

**Comment by Gankro** [comment]
I would like to propose stabilizing the likely/unlikely functions as-is to enable projects on stable ([like hashbrown](https://github.com/Amanieu/hashbrown/blob/master/src/raw/mod.rs#L12-L25)) to make use of these features where they can.

**Comment by gnzlbg** [comment]
> How do these intrinsics relate to the #[cold] attribute on functions?

`cold` is a calling convention on functions, while `likely`/`unlikely` annotate conditional jumps.

**Comment by hsivonen** [comment]
> If you want to apply this to `match`, `if let`, `while let`, etc

I'm not claiming that `likely()`/`unlikely()` would address those use cases. I'm saying that `likely()`/`unlikely()` would work for plain `if` and right now the search for perfection means we don't get to have prediction hints for plain `if`, either.

> Even for simple if-else code one cannot easily annotate the else branch using intrinsics:
> ```rust
> if foo { 
>    ...
> } else if ...  { ... }
> } else if ...  { ... }
> } else if ...  { ... }
> } core::intrinsics::unlikely(else) {  // ??
>    ...
> }
> ```

Since plain `if`conditions are evaluated as if the conditions were evaluated in order, the above should be:

```rust
if foo { 
   ...
} else if ...  { ... }
} else if ...  { ... }
} else if likely(...)  {
  ...
} else {
   ...
}
```

OTOH, what does the following mean:

```rust
if foo {
  ...
} else if bar {
  #![hot]
} else {
  ...
}
```

Does the middle annotation affect the prediction of `foo`, `bar` or both?

**Comment by pickfire** [comment]
Is not `match` supposed to be arranged in the order of likeliness? Same goes for branching.
```
match foo {
    A => ...,
    B => ...,
}
```
As well, is it just `likely` and `unlikely`? Or there maybe be something in middle?

**Comment by gnzlbg** [comment]
> Is not match supposed to be arranged in the order of likeliness?

No, it is not supposed to be arranged that way.

**Comment by kazimuth** [comment]
What work needs to be done to stabilize these? The RFC was merged, right?

**Comment by gnzlbg** [comment]
@kazimuth 

During the RFC discussion, the fact that these intrinsics work with expressions did not appear to be very important, but since these have been implemented, more people have wanted to use these with, e.g., `if let` and other control flow constructs like `match`. 

This has casted a bit of uncertainty on whether we should stabilize these intrinsics "as is", or whether solving this problem deserves a more general mechanism that works everywhere. As @Gankro mentioned, we can always add a more general mechanism later. From the few discussions about the more general feature in Zulip and Discord, it appears that a general isn't straight forward (e.g. for things like `match` arms some have expressed interest on being able to specify branch probabilities, as opposed to just "likely"/"unlikely").

**Comment by kazimuth** [comment]
Is someone willing to throw together a PR merging what we've got now?

**Comment by elichai** [comment]
I personally think that `likely`/`unlikely` (hot/cold) cover 99% of the use cases.
even in match.
I do think that it's important to be able to use it in `match` and `if let` etc.
otherwise it might get us backwards in terms of idiomatic rust.

**Comment by CreepySkeleton** [comment]
> even in match

How so? I mean, if a match guard is in use than yes
```rust
match val {
    x if likely(val > 5) => {},
    x => {}
}
```

but there's no way to apply them to a match statement  in "no guard" case

**Comment by elichai** [comment]
@CreepySkeleton As i've said a line later that I do think we should fix:
> I do think that it's important to be able to use it in `match` and `if let` etc.
> otherwise it might get us backwards in terms of idiomatic rust.

But I don't think that exact percentages are really that useful in practice, and being able to mark one of the match options as `likely` or `unlikely` is still very useful.

**Comment by Pzixel** [comment]
Upwote @gnzlbg 's post. We hardly need to decorate if let and pattern matching branches. I currently got 20% in my code by just rearranging match expression. It feels very unnatural I have to mimic C++ approach and native idiomatic rust is not that performant.

**Comment by matklad** [comment]
(this is mostly a drive by observation from a non-expert, so don't take to seriously). 

My understanding is that one of the problems here is that we have a narrow interface (`fn likely(bool) -> bool`), and there's a clear desire to have a more general one. However, it seems to me that, even if we implement a more general attribute-based solution, it still makes sense to have `fn likely`, as that is the most convenient interface to use for the common case. That is, I find this to be the best we can get in terms of readability:

```rust
if hint::likely(condition) {
}
```

It also seems that a function would be easier to discover and teach, and a good place to point to a more general facility, should we have one.

**Comment by matklad** [comment]
Also, I can not resist suggesting my own color. We have macros in pattern positions as of recently, so we can use a macro syntactically:

```rust
if hint::likely!(condition) {
}
match x {
    hint::likely!(Some(x)) => (),
    hitn::unlikely!(None) => (),
}
```

This should also provide syntactic affordance for later extensions

```rust
match x {
    hint::likely!(1, probability=0.5) => (),
    hint::likely!(2, probability=0.2) => (),
    hint::likely!(3, probability=0.2) => (),
    _ => (),
}
```

**Comment by pickfire** [comment]
Is there some sort of tailored optimizations that count the probability of match arms? IIRC rust can build some software with tailored optimization after a few runs, will that change the probability here? Or does that automatically make out these likely/unlikely intrinsic by itself?

**Comment by SimonSapin** [comment]
@matklad Almost all macros expand to Rust code. (I think `macro_rules` and `asm` are the two special cases?) `likely` is about providing information to the compiler that doesn’t have an "expanded" Rust code representation, so it fits an attribute much better in my opinion. Attributes also allow extensions like `#[likely(probability = 0.5)]`.

@pickfire It sounds like you’re looking for [Profile Guided Optimization](https://doc.rust-lang.org/rustc/profile-guided-optimization.html). It can be useful as well, but it’s not completely a replacement for source annotations since it’s harder to deploy and not deterministic.

**Comment by howard0su** [comment]
Do we consider the standard error handling in Rust like
match result  {
     Ok(xyz) => ....,
     Err(_) => ...., 
}
In such case, we can safely assume Ok() path is more common. And some applies to Option<> as well. Will this benefit the performance?

**Comment by Pzixel** [comment]
No, we can't. Rustc has no idea what `Ok` or `Err` is

**Comment by CryZe** [comment]
Similar to the attributes on match cases, the attributes could be applied to enum variants, so it would be
```rust
enum Result<T, E> {
    #[likely]
    Ok(T),
    Err(E),
}
```

**Comment by jgarvin** [comment]
C compilers have to rely on heuristics to try and guess the error case, the cool thing about Rust's more expressive type system is that Err is a strong indicator you're in an unlikely case and is used universally. The combination of being able to say "this variant itself is unlikely" and "this case is unlikely" is you can use the declaration on the variant to get 99% of the annotations you would otherwise need out of the way, then you can use the ones on specific matches to fine tune. There's even the potential to use likely/unlikely on variants to change object representation -- e.g. put bits distinguishing the common variants next to hot fields, and put bits for unlikely cases next to cold fields (maybe even in heap memory if the variants all already contain `Box<T>`).

**Comment by howard0su** [comment]
Can we extend [cold] attribute to able to annotate a **enum**. So that when we parse match or "if let", we can integrate this type of information via insert unlikely into the IR.

**Comment by EkremDincel** [comment]
Did anybody considered something like [this](https://godbolt.org/z/Pxv6cW) for annotating `match` statements:
```rust
fn get_some_with_the_high_possibilty_of_success_or_return_zero(option: Option<i32>) -> i32 {
    likely(matches!(option, Some(_)));
    match option {
        Some(value) => value,
        None => 0,
    }
}
```

Ideally, LLVM would optimize the `likely(matches!(option, Some(_)))` to no-op since it don't have any side effects. Would this be a reasonable and useful way of doing this and is this enough to hint the compiler for optimizations?

**Comment by bakape** [comment]
Code that never runs can already be optimized out with unreachable and
unreachable_unchecked.

**Comment by Kogia-sima** [comment]
LLVM properly optimizes combination of `likely`/`unlikely` and `.unwrap()` as expected.

https://godbolt.org/z/7Y6744

**Comment by EkremDincel** [comment]
@bakape this is not the point, I ask whether my example code hints the compiler about the likely of `Some(value) => ...` arm.

**Comment by CryZe** [comment]
@EkremDincel that doesn't seem to work as a hint

**Comment by EkremDincel** [comment]
@CryZe You are right: https://godbolt.org/z/ec6Mn8

**Comment by gilescope** [comment]
Am all for likely/unlikely - I think it makes it clearer when you're reading the code as to what's going on. Would love to see likely on `if let`, but I think it's a pretty widespread (apparently false) assumption that the first match statement is the most likely. Maybe we should find a way to make that assumption true so that the top match arm is the most likely?

**Comment by joshtriplett** [comment]
Nominating for libs-api: shall we approve moving this to `std::hint`?

**Comment by joshtriplett** [comment]
We discussed this in today's @rust-lang/lang meeting. We consider this done from a lang perspective, and it's up to T-libs-api to give this a final name (perhaps in `std::hint`) and stabilize this.

**Comment by Amanieu** [comment]
I personally feel that a language-level `#[likely]` attribute would be a better fit. It would work on `if`, `if let`, `match` arms, etc.

**Comment by joshtriplett** [comment]
@Amanieu I'd personally like to have both, since `likely` and `unlikely` intrinsics work in more contexts (e.g. sub-expressions).

**Comment by scottmcm** [comment]
@Amanieu I agree.  One thing that got mentioned in the meeting was that we have `#[cold]` for functions and closures, so allowing that on *blocks* might be an interesting way to phrase it.

(Albeit one can mostly write `#[cold] { ... }` as `(#[cold]||{ ... })()`.)

**Comment by Amanieu** [comment]
Clang's [`CodeGenFunction::EmitBranchOnBoolExpr`](https://github.com/llvm/llvm-project/blob/58ebc79a64a5ca8dacf504f55d904e030effa164/clang/lib/CodeGen/CodeGenFunction.cpp#L1662) has special handling for `__builtin_expect` (which is what `likely`/`unlikely` use under the hood) which is not present in rustc.

Specifically, it performs the following transformations with boolean expressions:
```
likely(a && b) => likely(a) && likely(b)
unlikely(a && b) => a && unlikely(b)
likely(a || b) => a || likely(b)
unlikely(a || b) => unlikely(a) || unlikely(b)
```

In rustc this would need to be done during THIR to MIR lowering, which would require the compiler to treat these intrinsics specially.

Additionally, the `llvm.expect` intrinsic is designed to work when its result is directly passed to a `br` LLVM IR instruction. It does not work if used outside of this context ([example](https://github.com/rust-lang/rust/issues/26179#issuecomment-787460803)), which is not very obvious.

**Comment by joshtriplett** [comment]
@Amanieu Those transformations seem potentially useful. I personally wouldn't consider them *necessary* to make these intrinsics useful, though. Would you consider those transformations to be blockers?

**Comment by iago-lito** [comment]
I have a naive question here: besides `likely` and `unlikely` hints, sometimes I have a guarantee that the branch taken will actually be (uniformly) `random`. Would such a hint be useful at all for compilation? Would it be equivalent to every branch being `unlikely`?

**Comment by the8472** [comment]
>  sometimes I have a guarantee that the branch taken will actually be (uniformly) random.

LLVM already has [unpreditable](https://llvm.org/docs/LangRef.html#unpredictable-metadata) for that. This came up in https://github.com/rust-lang/rust/issues/53823

**Comment by MSxDOS** [comment]
These intrinsics no longer have any effect for my code (starting with https://github.com/rust-lang/rust/pull/87570 ?).
Easily compared by importing them as `use core::intrinsics::{likely as unlikely, unlikely as likely};` or replacing with blank functions, in all cases produced binaries are identical.

**Comment by the8472** [comment]
@MSxDOS that's probably worth a regression issue.

**Comment by Amanieu** [comment]
We discussed this during the libs-api meeting and don't think that a function is the correct interface for exposing probabilities on LLVM IR branch instruction (or equivalents in other backends). We would instead prefer an attribute-based solution which would work on both `if` and `match` arms.

Passing this back to T-lang.

**Comment by joshtriplett** [comment]
For this tracking issue, for the intrinsics, we're marking these as "perma-unstable", like most LLVM intrinsics.

We'd love to see a lang MCP for `#[likely]` and `#[unlikely]` attributes.

**Comment by saethlin** [comment]
https://github.com/rust-lang/rust/pull/120370 should now work, at least sometimes. I don't know if https://github.com/rust-lang/rust/issues/88767 is fixed, but since reviewing the fix PR I now have the opinion that a `#[cold]` on arms or the `intrinsics::cold_path()` strategy is the right way to provide this branch hinting information.

**Comment by Clipi-12** [comment]
Proposal to reexport `intrinsics::cold_path()` in `std::hint` in [libs-team#510](https://github.com/rust-lang/libs-team/issues/510)

**Comment by tgross35** [comment]
We now have `likely`, `unlikely` and `cold_path` in `core::hint`. The feature currently points here but I think it's better to keep the library discussion separate from this issue's meta discussion (happy almost tenth birthday to this issue), so I'm moving the gates to https://github.com/rust-lang/rust/issues/136873.
